### ISSUE TITLE
Stricter typecasting option / prevent this implicit cast

### ISSUE THREAD

**Issue by Nielsbishere** [microsoft/DirectXShaderCompiler#4613]
It seems like the following is allowed:
```hlsl
void myFunc(uint x, float4x4 y);
myFunc(123, vp);
myFunc(vp, 123);
```
Where 123 could also just be a uint value.
Both would work, but only one would be right. I request a stricter implicit cast option or a fix for these casts since they're two indirections (cast from uint to float to float4x4 I'm guessing and the opposite as well). 
This could also be useful to implicit casts between vectors as well, as detecting this as a compiler warning could give our shader devs more useful info about possible issues instead of silently failing.

**Comment by Nielsbishere** [comment]
Another example of this is truncation from uint to float3. uint -> float should already give warnings because a float can only hold a 23-bit number. Example:
```hlsl
float3 test(float3 a) {
    return a;
}

float4 PSMain() : SV_TARGET {
	return float4(test((uint)0), 1);
}
```
This is really annoying that this doesn't generate a warning. The same is true when you truncate from float3 to uint. These are not safe casts!
```hlsl
float3 test(uint a) {
    return a.xxx;
}

float4 PSMain() : SV_TARGET {
	return float4(test(1.5.xxx), 1);
}
```

**Comment by damyanp** [comment]
Moving to HLSL specs to track tightening up implicit conversations in future language versions.

Clang will have warnings for these.  note that DXC will generate diagnostics for some of these cases if `-Wconversions` is passed.
